lib/semantics/type.cc

```diff
@@ -38,22 +39,253 @@ void DerivedTypeSpec::ReplaceScope(const Scope &scope) {
   scope_ = &scope;
 }
 
-ParamValue &DerivedTypeSpec::AddParamValue(
-    SourceName name, ParamValue &&value) {
+void DerivedTypeSpec::AddRawParamValue(
+    const std::optional<parser::Keyword> &keyword, ParamValue &&value) {
+  CHECK(parameters_.empty());
+  rawParameters_.emplace_back(keyword ? &*keyword : nullptr, std::move(value));
+}
+
+void DerivedTypeSpec::CookParameters(evaluate::FoldingContext &foldingContext) {
+  if (cooked_) {
+    return;
+  }
+  cooked_ = true;
+  auto &messages{foldingContext.messages()};
+  if (IsForwardReferenced()) {
+    messages.Say(typeSymbol_.name(),
+        "Derived type '%s' was used but never defined"_err_en_US,
+        typeSymbol_.name());
+    return;
+  }
+
+  // Parameters of the most deeply nested "base class" come first when the
+  // derived type is an extension.
+  auto parameterNames{OrderParameterNames(typeSymbol_)};
+  auto parameterDecls{OrderParameterDeclarations(typeSymbol_)};
+  auto nextNameIter{parameterNames.begin()};
+  RawParameters raw{std::move(rawParameters_)};
+  for (auto &[maybeKeyword, value] : raw) {
+    SourceName name;
+    common::TypeParamAttr attr{common::TypeParamAttr::Kind};
+    if (maybeKeyword) {
+      name = maybeKeyword->v.source;
+      auto it{std::find_if(parameterDecls.begin(), parameterDecls.end(),
+          [&](const Symbol &symbol) { return symbol.name() == name; })};
+      if (it == parameterDecls.end()) {
+        messages.Say(name,
+            "'%s' is not the name of a parameter for derived type '%s'"_err_en_US,
+            name, typeSymbol_.name());
+      } else {
+        // Resolve the keyword's symbol
+        maybeKeyword->v.symbol = const_cast<Symbol *>(&it->get());
+        attr = it->get().get<TypeParamDetails>().attr();
+      }
+    } else if (nextNameIter != parameterNames.end()) {
+      name = *nextNameIter++;
+      auto it{std::find_if(parameterDecls.begin(), parameterDecls.end(),
+          [&](const Symbol &symbol) { return symbol.name() == name; })};
+      CHECK(it != parameterDecls.end());
+      attr = it->get().get<TypeParamDetails>().attr();
+    } else {
+      messages.Say(name_,
+          "Too many type parameters given for derived type '%s'"_err_en_US,
+          typeSymbol_.name());
+      break;
+    }
+    if (FindParameter(name)) {
+      messages.Say(name_,
+          "Multiple values given for type parameter '%s'"_err_en_US, name);
+    } else {
+      value.set_attr(attr);
+      AddParamValue(name, std::move(value));
+    }
+  }
+}
+
+void DerivedTypeSpec::EvaluateParameters(
+    evaluate::FoldingContext &foldingContext) {
+  if (evaluated_) {
+    return;
+  }
+  evaluated_ = true;
```

Comment:
Is it necessary that `CookParameters` has been called so that `FindParameter` works correctly? If so, it might be a good idea to assert that here.

Reply:
Or I could call it.